README
NAME
    Text::MeCab - Alternate Interface To libmecab

SYNOPSIS
      use Text::MeCab;
      my $mecab = Text::MeCab->new({
        rcfile             => $rcfile,
        dicdir             => $dicdir,
        userdic            => $userdic,
        lattice_level      => $lattice_level,
        all_morphs         => $all_morphs,
        output_format_type => $output_format_type,
        partial            => $partial,
        node_format        => $node_format,
        unk_format         => $unk_format,
        bos_format         => $bos_format,
        eos_format         => $eos_format,
        input_buffer_size  => $input_buffer_size,
        allocate_sentence  => $allocate_sentence,
        nbest              => $nbest,
        theta              => $theta,
      });

      for (my $node = $mecab->parse($text); $node; $node = $node->next) {
         # See perdoc for Text::MeCab::Node for list of methods
         print $node->surface, "\n";
      }

      # use constants
      use Text::MeCab qw(:all);
      use Text::MeCab qw(MECAB_NOR_NODE);

      # want to use a command line arguments?
      my $mecab = Text::MeCab->new("--userdic=/foo/bar/baz", "-P");

      # check what mecab version we compiled against?
      print "Compiled with ", &Text::MeCab::MECAB_VERSION, "\n";

DESCRIPTION
    libmecab (http://mecab.sourceforge.ne.jp) already has a perl interface
    built with it, so why a new module? I just feel that while a subtle
    difference, making the perl interface through a tied hash is just...
    weird.

    So Text::MeCab gives you a more natural, Perl-ish way to access
    libmecab!

    WARNING: Please note that this module is primarily targetted for
    libmecab >= 0.90, so if things seem to be broken and your libmecab
    version is below 0.90, then you might want to consider upgrading
    libmecab first.

Text::MeCab AND SCOPING
    [NOTE: The memory management issue has been changed since 0.09]

    libmecab's default behavior is such that when you analyze a text and get
    a node back, that node is tied to the mecab "tagger" object that
    performed the analysis. Therefore, when that tagger is destroyed via
    mecab_destroy(), all nodes that are associated to it are freed as well.

    Text::MeCab defaults to the same behavior, so the following won't work:

      sub get_mecab_node {
         my $mecab = Text::MeCab->new;
         my $node  = $mecab->parse($_[0]);
         return $node;
      }

      my $node = get_mecab_node($text);

    By the time get_mecab_node() returns, the Text::MeCab object is
    DESTROY'ed, and so is $node (actually, the object exists, but it will
    complain when you try to access the node's internals, because the C
    struct that was there has already been freed).

    In such cases, use the dclone() method. This will copy the *entire* node
    structure and create a new Text::MeCab::Node::Cloned instance.

      sub get_mecab_node {
         my $mecab = Text::MeCab->new;
         my $node  = $mecab->parse($_[0]);
         return $node->dclone();
      }

    The returned Text::MeCab::Node::Cloned object is exactly the same as
    Text::MeCab::Node object on the surface. It just uses a different but
    very similar C struct underneath. It is blessed into a different
    namespace only because we need to use a different memory management
    strategy.

    Do be aware of the memory issue. You WILL use up twice as much memory.

    Also please note that if you try the first example, accessing the node
    *WILL* result in a segfault. This is *NOT* a bug: it's a feature :)
    While it is possible to control the memory management such that
    accessing a field in a node that has already expired results in a legal
    croak(), we do not go to the length to ensure this, because it will
    result in a performance penalty.

    Just remember that unless you dclone() a node, then you are NOT allowed
    to access it when the original tagger goes out scope:

       {
           my $mecab = Text::MeCab->new;
           $node = $mecab->parse(...);
       }

       $node->surface; # segfault!!!!

    Always remember to dclone() before doing this!

METHODS
  new HASHREF | LIST
    Creates a new Text::MeCab instance.

    You can either specify a hashref and use named parameters, or you can
    use the exact command line arguments that the mecab command accepts.

    Below is the list of accepted named options. See the man page for mecab
    for details about each option.

    rcfile
    dicdir
    lattice_level
    all_morphs
    output_format_type
    partial
    node_format
    unk_format
    bos_format
    eos_format
    input_buffer_size
    allocate_sentence
    nbest
    theta

  parse SCALAR
    Parses the given text via mecab, and returns a Text::MeCab::Node object.

SEE ALSO
    http://mecab.sourceforge.ne.jp

AUTHOR
    (c) 2006 Daisuke Maki <dmaki@cpan.org> All rights reserved.



Changes
Changes
=======

0.20006 07 Mar 2008
  - Makefile.PL tweak for Win32. Patched by Kenichi Ishigaki
  - We've also been reported that older libmecab doesn't work with Text::MeCab.
    This is a known issue mainly caused by the fact that I have no access to
    older libmecab in my dev environment. If you know of particular combo
    (libmecab version against a particular feature) that doesn't work, please
    send in patches and reports so we can fix them in future releases

0.20005 06 Mar 2008
  - Now we use sv_setref_pv instead of sv_bless and such. This fixes
    a major leakage that prevented Text::MeCab::Node objects from being
    garbage collected until global destruction time. I have no other
    explanation other than that once I switched to sv_setref_pv, everything
    just worked.
  
    Upgrade from previous 0.2000x versions STRONGLY recommended.

0.20004 10 Jan 2008
  - Text::MeCab::Dict has now been confirmed working with mecab-ipadic-20070801
    * tweak the encoding
    * don't use Text::CSV_XS

0.20003 10 Jan 2008
  - Of course, it's always a good idea to *actually* include the new
    module. Yikes. If you downloaded 0.20002 by some odd chance, please 
    use this release instead.

0.20002 10 Jan 2008
  - Add Text::MeCab::Dict, which is a simple wrapper to work with 
    mecab dictionary. Only supports ipadic.

0.20001 09 Jan 2008
  - Properly use ExtUtils::MakeMaker::prompt() to ask interactive
    questions. Pointed out by David Cantrell.

0.20000 08 Jan 2008
  - No code change. Releasing.

0.20000_01 07 Jan 2008
  - Complete rewrite
    * work with libmecab 0.96
    * rework tests
    * use typemaps wisely
    * default encoding is now utf-8
    * now require Encode

0.17 01 May 2007
  - Fixup stupid Copy() problem.

0.16 16 Apr 2007
  - Change the internal C structure to be thin wrappers around mecab_node_t
  - Implement a node->format() method

0.15 29 Jan 2007
  - 0.14 had upload problems. Repackage.

0.14 28 Jan 2007
  - Fix argument passing to mecab_new(). Reported by Naoki Tomita.

0.13 08 Aug 2006
  - Fix tools/probe_mecab.pl so that there are not spurfulous whitespaces
    around the flags (http://d.hatena.ne.jp/t-tkzw/20060730/p2).

0.12 15 Jul 2006
  - Apply "Poor Puppy" patch from Kenichi Ishigaki (charsbar)
  - Properly ask for the dictionary encoding when running perl Build.PL.
    This will create t/strings.dat with that encoding.

0.11 14 Jul 2006
  - Silence more warnings
  - Force use of -Wall at compilation time

0.10 14 Jul 2006
  - remove spurfuluous parse_wakati.pl
  - remove debug statements
  - silence warnings

0.09 12 Jul 2006
  - "Hey, it's the day before my birthday, but I'm releasing a new module" release
  - Switch default behavior of Text::MeCab when it goes out of scope.
    See "Text::MeCab AND SCOPE" section in Text::MeCab POD.
    This all prompted by post at http://d.hatena.ne.jp/t-tkzw/20060710/p1.
  - Add new Text::MeCab::Node::Cloned to workaround.
  - Add warning about not using cloned node when Text::MeCab goes out of
    scope.

0.08 - 05 Jul 2006
  - Apply suggestions by charsbar when prompting for compile/link options
    for Win32
    (and actually release the changes -- this has been sitting on my SVK
     client for a month)

0.07 - 09 Jun 2006
  - The way we were passing arguments to mecab_new() was totally wrong.
    Fixed.

0.06 - 08 May 2006
  - Require ExtUtils::MakeMaker >= 6.25 to avoid Build.PL being
    executed after Makefile.PL
  - Fix INIT -> PREINIT (reported by charsbar)
  - Fix how XSRETURN_UNDEF was working (reported by charsbar)
  - Fix tools/probe_mecab.pl (reported by charsbar)
  - Use ppport.h.

0.05 - 04 May 2006
  - Fix typos
  - Add tests to MANIFEST. argh.
  - Actually test against libmecab < 0.90. Now tests pass.
  - Bump up version to 0.05 for historical reasons. There was another
    version of Text::MeCab that MAKAMAKA had written which went up to
    0.04.

0.02_03 - 04 May 2006
  - Correct reference counting for prev().
  - Add tests for detatched Text::MeCab::Node.
  - Tweak docs

0.02_02 - 04 May 2006
  - Fix tools/probe_mecab.pl such that it prompts the user for some
    required parameters when installing on Windows.
  - Try fixing Makefile.PL once again.
  - Explicitly make copies of mecab_node_t in the XS, so that you
    can now manipulate the nodes even *after* your instance of
    Text::MeCab has gone away

0.02_01 - 03 May 2006
  - Hey, I didn't know people were going to jump on to this module,
    seriously... 

  - Fix building when using Makefile.PL instead of Build.PL
    (refactor important bits to tools/probe_mecab.pl).
  - Attempt to work with MeCab < 0.90.
  - Move benchmark.pl to tools/benchmark.pl.
  - Fix problem caused by Text::MeCab->new() (no parameters).
  - Add Text::MeCab::MECAB_VERSION to display the mecab version
    we compiled against.

  TODO:
  - Try to at least give out a warning when executing a code like this:

    my $node;
    {
        my $mecab = Text::MeCab->new;
        $mecab->parse("......");
        $mecab = undef;
    }

    for(; $node; $node = $node->next) {
       print $node->surface, "\n";
    }

    I'm having a hard time detecting when a node is deallocated, though.
    we shall see.
    
0.02 - 02 May 2006
  - Accept command line arguments as well as the named parameters in hashref
  - Fixed problem where constants MECAB_ weren't declared int he correct
    namespace.

0.01 - 02 May 2006
  - Initial release.

lib/Text/MeCab.pm
# $Id: /mirror/coderepos/lang/perl/Text-MeCab/trunk/lib/Text/MeCab.pm 43505 2008-03-09T11:09:55.267941Z daisuke  $
#
# Copyright (c) 2006-2008 Daisuke Maki <daisuke@endeworks.jp>
# All rights reserved.

package Text::MeCab;
use strict;
use warnings;
use 5.006;
use Exporter 'import';
our ($VERSION, @ISA, %EXPORT_TAGS, @EXPORT_OK);
BEGIN
{
    $VERSION = '0.20007';
    if ($] > 5.006) {
        require XSLoader;
        XSLoader::load(__PACKAGE__, $VERSION);
    } else {
        require DynaLoader;
        @ISA = qw(DynaLoader);
        __PACKAGE__->bootstrap;
    }

    %EXPORT_TAGS = (all => [ qw(MECAB_NOR_NODE MECAB_UNK_NODE MECAB_BOS_NODE MECAB_EOS_NODE) ]);
    @EXPORT_OK = map { @$_ } values %EXPORT_TAGS;
}

my %BOOLEAN_OPTIONS = (
    map { ($_, 'bool') } qw(
        --all-morphs --partial --allocate-sentence --version --help
    )
);

sub new
{
    my $class = shift;

    my %args = ref $_[0] eq 'HASH' ? %{$_[0]} : @_;
    $args{'allocate-sentence'} = 1;

    my @args = ('perl-TextMeCab');
    while (my($key, $value) = each %args) {
        $key =~ s/_/-/g;
        $key =~ s/^/--/;

        if (exists $BOOLEAN_OPTIONS{$key}) {
            push @args, $key;
        } else {
            push @args, join('=', $key, $value);
        }
    }

    $class->_XS_new(\@args);
}

1;

__END__

=head1 NAME

Text::MeCab - Alternate Interface To libmecab

=head1 SYNOPSIS

  use Text::MeCab;
  my $mecab = Text::MeCab->new({
    rcfile             => $rcfile,
    dicdir             => $dicdir,
    userdic            => $userdic,
    lattice_level      => $lattice_level,
    all_morphs         => $all_morphs,
    output_format_type => $output_format_type,
    partial            => $partial,
    node_format        => $node_format,
    unk_format         => $unk_format,
    bos_format         => $bos_format,
    eos_format         => $eos_format,
    input_buffer_size  => $input_buffer_size,
    allocate_sentence  => $allocate_sentence,
    nbest              => $nbest,
    theta              => $theta,
  });

  for (my $node = $mecab->parse($text); $node; $node = $node->next) {
     # See perdoc for Text::MeCab::Node for list of methods
     print $node->surface, "\n";
  }

  # use constants
  use Text::MeCab qw(:all);
  use Text::MeCab qw(MECAB_NOR_NODE);

  # check what mecab version we compiled against?
  print "Compiled with ", &Text::MeCab::MECAB_VERSION, "\n";

=head1 DESCRIPTION

libmecab (http://mecab.sourceforge.ne.jp) already has a perl interface built 
with it, so why a new module? I just feel that while a subtle difference,
making the perl interface through a tied hash is just... weird.

So Text::MeCab gives you a more natural, Perl-ish way to access libmecab!

WARNING: Version 0.20000 has only been tested against libmecab 0.96.

=head1 Text::MeCab AND FORMATS

mecab allows users to specify an output format, via --*-format options.
These are respected ONLY if you use the format() method:

  my $mecab = Text::MeCab->new({
    output_format_type => "user",
    node_format => "%m %pn"
  });

  for(my $node = $mecab->parse($text); $node; $node = $node->next) {
    print $node->format($mecab);
  }

Note that you also need to set the output_format_type parameter as well.

=head1 Text::MeCab AND SCOPING

[NOTE: The memory management issue has been changed since 0.09]

libmecab's default behavior is such that when you analyze a text and get a
node back, that node is tied to the mecab "tagger" object that performed the
analysis. Therefore, when that tagger is destroyed via mecab_destroy(),
all nodes that are associated to it are freed as well.

Text::MeCab defaults to the same behavior, so the following won't work:

  sub get_mecab_node {
     my $mecab = Text::MeCab->new;
     my $node  = $mecab->parse($_[0]);
     return $node;
  }

  my $node = get_mecab_node($text);

By the time get_mecab_node() returns, the Text::MeCab object is DESTROY'ed, 
and so is $node (actually, the object exists, but it will complain when you
try to access the node's internals, because the C struct that was there
has already been freed).

In such cases, use the dclone() method. This will copy the *entire* node
structure and create a new Text::MeCab::Node::Cloned instance. 

  sub get_mecab_node {
     my $mecab = Text::MeCab->new;
     my $node  = $mecab->parse($_[0]);
     return $node->dclone();
  }

The returned Text::MeCab::Node::Cloned object is exactly the same as 
Text::MeCab::Node object on the surface. It just uses a different but
very similar C struct underneath. It is blessed into a different namespace
only because we need to use a different memory management strategy.

Do be aware of the memory issue. You WILL use up twice as much memory.

Also please note that if you try the first example, accessing the node 
*WILL* result in a segfault. This is *NOT* a bug: it's a feature :) 
While it is possible to control the memory management such that accessing 
a field in a node that has already expired results in a legal croak(), 
we do not go to the length to ensure this, because it will result in
a performance penalty. 

Just remember that unless you dclone() a node, then you are NOT allowed to
access it when the original tagger goes out scope:

   {
       my $mecab = Text::MeCab->new;
       $node = $mecab->parse(...);
   }

   $node->surface; # segfault!!!!

Always remember to dclone() before doing this!

=head1 PERFORMANCE

Belows is the result of running tools/benchmark.pl on my PowerBook:

  daisuke@beefcake Text-MeCab$ perl tools/benchmark.pl 
               Rate      mecab text_mecab
  mecab      5.53/s         --       -63%
  text_mecab 14.9/s       170%         --

=head1 METHODS

=head2 new HASHREF | LIST

Creates a new Text::MeCab instance.

You can either specify a hashref and use named parameters, or you can use the
exact command line arguments that the mecab command accepts.

Below is the list of accepted named options. See the man page for mecab for 
details about each option.

=over 4

=item B<rcfile>

=item B<dicdir>

=item B<lattice_level>

=item B<all_morphs>

=item B<output_format_type>

=item B<partial>

=item B<node_format>

=item B<unk_format>

=item B<bos_format>

=item B<eos_format>

=item B<input_buffer_size>

=item B<allocate_sentence>

=item B<nbest>

=item B<theta>

=back

=head2 parse SCALAR

Parses the given text via mecab, and returns a Text::MeCab::Node object.

=head2 ENCODING

  my $encoding = Text::MeCab::ENCODING

Returns the encoding of the underlying mecab library that was detected at
compile time.

=head2 MECAB_VERSION

The version number from libmecab's mecab_version()

=head2 MECAB_TARGET_VERSION

=head2 MECAB_TARGET_MAJOR_VERSION

=head2 MECAB_TARGET_MINOR_VERSION

The version number detected at compile time of Text::MeCab. 

=head2 MECAB_NOR_NODE

=head2 MECAB_UNK_NODE

=head2 MECAB_BOS_NODE

=head2 MECAB_EOS_NODE

=head2 MECAB_USR_DIC

=head2 MECAB_SYS_DIC

=head2 MECAB_UNK_DIC

=head2 MECAB_CONFIG

Path to mecab-config, if available.

=head1 SEE ALSO

http://mecab.sourceforge.ne.jp

=head1 LICENSE

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See http://www.perl.com/perl/misc/Artistic.html

=head1 AUTHOR

Copyright (c) 2006-2008 Daisuke Maki E<lt>daisuke@endeworks.jpE<gt>
All rights reserved.

=cut
